### PR TITLE
feat: add conversion of softmax from onnx to tosa

### PR DESCRIPTION
**Pull request by maxbartel**

### CODE REVIEW

File: src/Conversion/ONNXToTOSA/Math/Softmax.cpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+    // reduce_sum on last dimension
+    int32_t inputRank = inputType.getShape().size();
+
+    SmallVector<int64_t> rsumShapeV(
+        inputType.getShape().begin(), inputType.getShape().end());
+    rsumShapeV[inputRank - 1] = 1;
+    ArrayRef<int64_t> rsumShape(rsumShapeV);
+
+    // Floating-point loewring is more direct:
```
Comment (flemairen6) [inline src/Conversion/ONNXToTOSA/Math/Softmax.cpp:54]:
```suggestion
    // Floating-point lowering is more direct:
```

File: src/Conversion/ONNXToTOSA/Math/Softmax.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+namespace onnx_mlir {
+
+namespace {
+
+class ONNXSoftmaxLoweringToTOSA : public OpConversionPattern<ONNXSoftmaxOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+  using OpAdaptor = typename ONNXSoftmaxOp::Adaptor;
+  LogicalResult matchAndRewrite(ONNXSoftmaxOp op, OpAdaptor adaptor,
```
Comment (flemairen6) [inline src/Conversion/ONNXToTOSA/Math/Softmax.cpp:32]: What happens when the `axis` attribute is set? Is it not supported by TOSA?

Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/Math/Softmax.cpp:32]: I missed that. Added this in [550cc52](https://github.com/ljfitz/onnx-mlir/pull/65/commits/550cc52a6963ff85ac934bbc3bd674ee795f0e38)

**Review by flemairen6 — APPROVED** [review]

File: test/mlir/conversion/onnx_to_tosa/Math/Softmax.mlir [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+// CHECK: %[[VAR3:.*]] = "tosa.mul"(%[[VAR0]], %[[VAR2]]) {shift = 0 : i32}
+}
+
+// -----
+func.func @test_axis_zero_softmax(%arg0: tensor<13x21x3xf32>) -> tensor<13x21x3xf32> {
+  %2 = "onnx.Softmax"(%arg0) {axis = 1 : si64} : (tensor<13x21x3xf32>) -> tensor<13x21x3xf32>
+  func.return %2 : tensor<13x21x3xf32>
+// CHECK-LABEL: test_axis_zero_softmax
+// CHECK-DAG: %[[VAR0:.*]] = "tosa.exp"(%arg0)
+// CHECK-DAG: %[[VAR1:.*]] = "tosa.reduce_sum"(%[[VAR0]]) {axis = 1 : i64} : (tensor<13x21x3xf32>) -> tensor<13x?x?xf32>
```
Comment (ljfitz) [inline test/mlir/conversion/onnx_to_tosa/Math/Softmax.mlir:19]: Does the `tosa::CreateOpAndInfer` not manage to determine the static shape of the result type?

File: test/mlir/conversion/onnx_to_tosa/Math/Softmax.mlir [hunk 4]
```diff
@@ -0,0 +1,22 @@
+// RUN: onnx-mlir-opt --shape-inference --convert-onnx-to-tosa %s -split-input-file | FileCheck %s
+
+func.func @test_softmax(%arg0: tensor<13x21x3xf32>) -> tensor<13x21x3xf32> {
+  %2 = "onnx.Softmax"(%arg0)   : (tensor<13x21x3xf32>) -> tensor<13x21x3xf32>
+  func.return %2 : tensor<13x21x3xf32>
+// CHECK-LABEL: test_softmax
+// CHECK-DAG: %[[VAR0:.*]] = "tosa.exp"(%arg0)
+// CHECK-DAG: %[[VAR1:.*]] = "tosa.reduce_sum"(%[[VAR0]]) {axis = 2 : i64}
+// CHECK-DAG: %[[VAR2:.*]] = "tosa.reciprocal"(%[[VAR1]])
+// CHECK: %[[VAR3:.*]] = "tosa.mul"(%[[VAR0]], %[[VAR2]]) {shift = 0 : i32}
```
Comment (ljfitz) [inline test/mlir/conversion/onnx_to_tosa/Math/Softmax.mlir:10]: As we are quite static shape oriented at present, please include the types for ease of review.

**Review by ljfitz — CHANGES_REQUESTED** [review]
Minor comments, but I'd like a chance to do another review when you've made them in case something else comes to light when I re-read.

File: test/mlir/conversion/onnx_to_tosa/Math/Softmax.mlir (hunk 4 above)
Comment (maxbartel) [reply test/mlir/conversion/onnx_to_tosa/Math/Softmax.mlir:10]: That was an error when creating the` rsumShape`. The shape handled only reducing in the last dimension, the integer attribute could mark all axis. The builder must have noticed the mismatch and inserted dynamic shapes.

File: test/mlir/conversion/onnx_to_tosa/Math/Softmax.mlir (hunk 3 above)
Comment (maxbartel) [reply test/mlir/conversion/onnx_to_tosa/Math/Softmax.mlir:19]: This was the shape bug mentioned

File: src/Conversion/ONNXToTOSA/Math/Softmax.cpp [hunk 5]
```diff
@@ -63,27 +74,17 @@ class ONNXSoftmaxLoweringToTOSA : public OpConversionPattern<ONNXSoftmaxOp> {
...
-        tosa::CreateOpAndInfer<tosa::MulOp>(rewriter, op->getLoc(), outputType,
-            op1ExpIn.getResult(), op3ReciprocalOp2.getResult(), 0)
-            .getResult();
-    if (!result)
-      return rewriter.notifyMatchFailure(op, "Legalization was not possible");
+    tosa::CreateReplaceOpAndInfer<tosa::MulOp>(rewriter, op, outputType,
+        op1ExpIn.getResult(), op3ReciprocalOp2.getResult(), 0);
 
-    rewriter.replaceOp(op, {result.value()});
+    // rewriter.replaceOp(op, {result});
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/Math/Softmax.cpp:87]:
```suggestion
```

File: test/mlir/conversion/onnx_to_tosa/Math/Softmax.mlir [hunk 6]
```diff
@@ -4,19 +4,19 @@ func.func @test_softmax(%arg0: tensor<13x21x3xf32>) -> tensor<13x21x3xf32> {
...
   func.return %2 : tensor<13x21x3xf32>
 // CHECK-LABEL: test_softmax
-// CHECK-DAG: %[[VAR0:.*]] = "tosa.exp"(%arg0)
-// CHECK-DAG: %[[VAR1:.*]] = "tosa.reduce_sum"(%[[VAR0]]) {axis = 2 : i64}
-// CHECK-DAG: %[[VAR2:.*]] = "tosa.reciprocal"(%[[VAR1]])
-// CHECK: %[[VAR3:.*]] = "tosa.mul"(%[[VAR0]], %[[VAR2]]) {shift = 0 : i32}
+// CHECK-DAG: %[[VAR0:.*]] = "tosa.exp"(%arg0) : (tensor<13x21x3xf32>) -> tensor<13x21x3xf32>
+// CHECK-DAG: %[[VAR1:.*]] = "tosa.reduce_sum"(%[[VAR0]]) {axis = 2 : i64} : (tensor<13x21x3xf32>) -> tensor<13x21x1xf32>
+// CHECK-DAG: %[[VAR2:.*]] = "tosa.reciprocal"(%[[VAR1]]) : (tensor<13x21x1xf32>) -> tensor<13x21x1xf32>
+// CHECK: %[[VAR3:.*]] = "tosa.mul"(%[[VAR0]], %[[VAR2]]) {shift = 0 : i32}  : (tensor<13x21x3xf32>, tensor<13x21x1xf32>) -> tensor<13x21x3xf32>
```
Comment (ljfitz) [inline test/mlir/conversion/onnx_to_tosa/Math/Softmax.mlir:10]: NIT: I don't see how these operations could appear in a different order due to the SSA dependencies, so the `-DAG` seems unnecessary, though not harmful, to me.

**Review by ljfitz — APPROVED** [review]

**Pull request merged**